src/utils/ddb/ddb_vos.c

```diff
@@ -1771,29 +1771,148 @@ sync_cb(struct ddbs_sync_info *info, void *cb_args)
+		if (!SUCCESS(rc)) {
+			D_ERROR("mark target ready failed: "DF_RC"\n", DP_RC(rc));
+			args->sync_rc = rc;
+			return;
+		} 
+	}
+
 	if (args->sync_complete_cb) {
 		rc = args->sync_complete_cb(args->sync_cb_args, pool_id,
-					    info->dsi_hdr->bbh_vos_id,
-					    info->dsi_hdr->bbh_blob_id,
-					    blob_size, info->dsi_dev_id);
+					    tgt_id, blob_id, blob_size, info->dsi_dev_id, st);
+	}
+}
+
+static void
+sync_cb_data(struct ddbs_sync_info *info, void *cb_args)
+{
+	do_sync_cb(info->dsi_hdr->bbh_pool, info->dsi_hdr->bbh_vos_id, 
+				 	info->dsi_hdr->bbh_blob_id, info->st, NULL, info, cb_args);
+}
+
+static struct smd_pool_info *
+find_pool_by_blobid_v1(struct meta_header_v1 *hdr, d_list_t pool_list, int pool_list_cnt)
+{
+	struct smd_pool_info    *pool_info = NULL;
+	struct smd_pool_info    *tmp = NULL;
+	int i = 0;
+	int found = 0;
+	int cnt = 0;
+
+	enum smd_dev_type st;
+	uint32_t target_id = hdr->mh_vos_id;
+	uint64_t blob_id[SMD_DEV_TYPE_MAX];
+	blob_id[SMD_DEV_TYPE_DATA] = hdr->mh_data_blobid;
+	blob_id[SMD_DEV_TYPE_META] = hdr->mh_meta_blobid;
+	blob_id[SMD_DEV_TYPE_WAL] = hdr->mh_wal_blobid;
+
+	d_list_for_each_entry_safe(pool_info, tmp, &(pool_list), spi_link) {
+		for (st = SMD_DEV_TYPE_DATA; st < SMD_DEV_TYPE_MAX; st++) {
+			for (i = 0; i < pool_info->spi_tgt_cnt[st]; i++) {
+				if (pool_info->spi_tgts[st][i] == target_id)
+					break;
+			}
+			if (pool_info->spi_blobs[st][i] != blob_id[st])
+				break;
+			found++;
+		}
+
+		// Found 
+		if (found == SMD_DEV_TYPE_MAX) 
+			return pool_info;
+		if (++cnt >= pool_list_cnt)
+			break;
+	}
+
+	return NULL;
+}
+
+static void
+sync_cb_meta(struct ddbs_sync_info *info, void *cb_args)
+{
+	struct dv_sync_cb_args	*args = cb_args;
+	struct smd_pool_info    *pool_info = NULL;
+	struct meta_header * hdr = info->dsi_meta_hdr;
+	uuid_t pool_id;
+
+    if (hdr->mh_version == 1) {
+		/* For Phase 1, No pool id in meta blob. get from Data Blob */
+		struct meta_header_v1 *v1 = (struct meta_header_v1 *)hdr;
+		pool_info = find_pool_by_blobid_v1(v1, args->pool_list, args->pool_list_cnt);
+		if (!pool_info) {
+			D_PRINT("Warn: Unable to find the pool with blobid "
+				"[d:%#lx][m:%#lx][w:%#lx] tgt_id %u\n", 
+				v1->mh_data_blobid, v1->mh_meta_blobid,
+				v1->mh_wal_blobid, v1->mh_vos_id);
+			args->sync_rc = -DER_UNKNOWN;
+			return;
+		}
+		uuid_copy(pool_id, pool_info->spi_id);
+	} else {
+		uuid_copy(pool_id, hdr->mh_pool_id);
+	}
+
+	// Sync Meta Blob
+	do_sync_cb(pool_id, info->dsi_meta_hdr->mh_vos_id, 
+				info->dsi_meta_hdr->mh_meta_blobid, SMD_DEV_TYPE_META,
+				pool_info, info, cb_args);
+
+	// Sync Wal Blob
+	do_sync_cb(pool_id, info->dsi_meta_hdr->mh_vos_id, 
+		info->dsi_meta_hdr->mh_wal_blobid, SMD_DEV_TYPE_WAL,
+		pool_info, info, cb_args);
+
+	return;
+}
+
+static void
+sync_cb_wal(struct ddbs_sync_info *info, void *cb_args)
+{
+	/**
+	 * Ignore Wal Blob. It will be sync in Meta Blob.
+	 * Because No BlobID in Wal Header.
+	 **/
```

Comment:
I think the blob ID could be obtained for any opened blob by spdk_blob_get_id(), right?

I'm wondering if we need to verify if the WAL blob is existing and the blob ID is same to the ID stored in meta header?